File: app/src/main/java/com/wa2c/android/medoly/plugin/action/lrclyrics/service/PluginReceivers.java
package com.wa2c.android.medoly.plugin.action.lrclyrics.service;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Build;
import android.preference.PreferenceManager;

import com.wa2c.android.medoly.library.MediaPluginIntent;
import com.wa2c.android.medoly.library.PluginOperationCategory;
import com.wa2c.android.medoly.library.PluginTypeCategory;
import com.wa2c.android.medoly.plugin.action.lrclyrics.R;
import com.wa2c.android.medoly.plugin.action.lrclyrics.util.Logger;

import org.androidannotations.api.builder.ServiceIntentBuilder;

/**
 * Execute receiver.
 */
public class PluginReceivers {

    public static abstract class AbstractPluginReceiver extends BroadcastReceiver {
        @Override
        public void onReceive(Context context, Intent intent) {
            Logger.d("onReceive: " + this.getClass().getSimpleName());

            MediaPluginIntent serviceIntent = new MediaPluginIntent(intent);
            Class c = this.getClass();
            serviceIntent.putExtra(AbstractPluginService.RECEIVED_CLASS_NAME, c.getName());

            SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
            ServiceIntentBuilder builder = null;
            if (c == EventGetLyricsReceiver.class || c == ExecuteGetLyricsReceiver.class) {
                String operation = pref.getString(context.getString(R.string.pref_event_get_lyrics), "");
                if (serviceIntent.hasCategory(PluginOperationCategory.OPERATION_EXECUTE) ||
                    (serviceIntent.hasCategory(PluginOperationCategory.OPERATION_MEDIA_OPEN) && PluginOperationCategory.OPERATION_MEDIA_OPEN.name().equals(operation)) ||
                    (serviceIntent.hasCategory(PluginOperationCategory.OPERATION_PLAY_START) && PluginOperationCategory.OPERATION_PLAY_START.name().equals(operation))) {
                    PluginGetLyricsService_.intent(context).stop();
                    serviceIntent.setClass(context, PluginGetLyricsService_.class);
                    builder = PluginGetLyricsService_.intent(context).process(serviceIntent);
                } else {
                    AbstractPluginService.sendResult(context, serviceIntent, null);
                }
            } else if (c == ExecuteSearchLyricsReceiver.class) {
                if (serviceIntent.hasCategory(PluginTypeCategory.TYPE_RUN)) {
                    serviceIntent.setClass(context, PluginRunService_.class);
                    PluginRunService_.intent(context).stop();
                    //PluginRunService_.intent(context).process(serviceIntent).start();
                    builder = PluginRunService_.intent(context).process(serviceIntent);
                }
            }

            if (builder != null) {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    context.startForegroundService(builder.get());
                } else {
                    context.startService(builder.get());
                }
            }
        }
    }

    // Event

    public static class EventGetLyricsReceiver extends AbstractPluginReceiver { }

    // Execution

    public static class ExecuteGetLyricsReceiver extends AbstractPluginReceiver { }

    public static class ExecuteSearchLyricsReceiver extends AbstractPluginReceiver { }

}

File: app/src/main/java/com/wa2c/android/medoly/plugin/action/lrclyrics/service/PluginReceivers.kt
package com.wa2c.android.medoly.plugin.action.lrclyrics.service

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.os.Build
import com.wa2c.android.medoly.library.*
import com.wa2c.android.medoly.plugin.action.lrclyrics.R
import com.wa2c.android.medoly.plugin.action.lrclyrics.util.AppUtils
import com.wa2c.android.prefs.Prefs
import timber.log.Timber

/**
 * Execute receiver.
 */
class PluginReceivers {

    abstract class AbstractPluginReceiver : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            Timber.d("onReceive: " + this.javaClass.simpleName)
            val result = receive(context, MediaPluginIntent(intent))
            setResult(result.resultCode, null, null)
        }

        /**
         * Receive data.
         */
        private fun receive(context: Context, pluginIntent: MediaPluginIntent): PluginBroadcastResult {
            var result =  PluginBroadcastResult.CANCEL

            val propertyData = pluginIntent.propertyData ?: return result
            val prefs = Prefs(context)

            if (this is EventGetLyricsReceiver ||
                    this is ExecuteGetLyricsReceiver) {
                // category
                if (!pluginIntent.hasCategory(PluginTypeCategory.TYPE_GET_LYRICS)) {
                    AppUtils.sendResult(context, pluginIntent)
                    return result
                }
                // media
                if (propertyData.isMediaEmpty) {
                    AppUtils.showToast(context, R.string.message_no_media)
                    AppUtils.sendResult(context, pluginIntent)
                    return result
                }
                // property
                if (propertyData.getFirst(MediaProperty.TITLE).isNullOrEmpty() || propertyData.getFirst(MediaProperty.ARTIST).isNullOrEmpty()) {
                    AppUtils.sendResult(context, pluginIntent)
                    return result
                }

                val operation = prefs.getString(R.string.pref_event_get_lyrics, defRes = R.string.pref_default_event_get_lyrics)
                if (!pluginIntent.hasCategory(PluginOperationCategory.OPERATION_EXECUTE) && !pluginIntent.hasCategory(operation)) {
                    AppUtils.sendResult(context, pluginIntent)
                    return result
                }

                // service
                pluginIntent.setClass(context, PluginGetLyricsService::class.java)
                result = PluginBroadcastResult.PROCESSING
            } else if (this is ExecuteSearchLyricsReceiver) {
                // category
                if (!pluginIntent.hasCategory(PluginTypeCategory.TYPE_RUN)) {
                    return result
                }

                // service
                pluginIntent.setClass(context, PluginRunService::class.java)
                result = PluginBroadcastResult.COMPLETE
            }

            pluginIntent.putExtra(AbstractPluginService.RECEIVED_CLASS_NAME, this.javaClass.name)
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                context.startForegroundService(pluginIntent)
            } else {
                context.startService(pluginIntent)
            }
            return result
        }
    }

    // Event

    class EventGetLyricsReceiver : AbstractPluginReceiver()

    // Execution

    class ExecuteGetLyricsReceiver : AbstractPluginReceiver()

    class ExecuteSearchLyricsReceiver : AbstractPluginReceiver()

}
